### ISSUE TITLE
Lambda TypeGuard false positive

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3719]
**Describe the bug**
```python
from typing import Callable

from typing_extensions import TypeGuard


def bad_foo() -> Callable[[object], TypeGuard[int]]:
    # Error reported here that Expression of type "(x: object) -> bool" 
    # cannot be assigned to return type "(object) -> TypeGuard[int]"
    return lambda x: isinstance(x, int) 

def good_foo() -> Callable[[object], TypeGuard[int]]: # This one works as expected.
    def _foo(x: object) -> TypeGuard[int]:
        return isinstance(x, int) 
    return _foo
```
**To Reproduce**
I reproduced both with cli and pylance.

**Expected behavior**
A lambda can be compatible with a typeguard.

**VS Code extension or command-line**
pyright 1.1.261

**Additional context**
Admittingly Mypy does give a similar error. And writing a small def instead of a lambda is an easy workaround so understandable if this is an edge case.

**Comment by erictraut** [comment]
Yeah, it won't work to use a lambda here. The type of a `TypeGuard` cannot be inferred. It needs to be explicitly annotated.

You should use a function rather than a lambda, since the lambda syntax doesn't allow you to provide an explicit return type annotation.
